### PR TITLE
Use Tensorflow's simple `LazyLoader` class to minimize import times of expensive packages

### LINKED ISSUE #4370: Slow imports
This is a non-urgent issue that I noticed during #4345 -- running `sct_deepseg -h` takes a noticeably long time. 

I'm curious as to why this is, and I would love to try to reduce the "startup" time for scripts if I can.

### PR DESCRIPTION
**Pull request by joshuacwnewton**
> [!NOTE]
> This PR is largely the same as #4490, however it swaps the use of `importlib`'s [`lazy_import`](https://docs.python.org/3/library/importlib.html#implementing-lazy-imports) for a copy of Tensorflow's `LazyLoader`, which I've found to perform much more predictably. I had been struggling with `lazy_import` for much of the day (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/57e3bf272e7be5eccf03322f63bedcaa8f7df86c), trying to track down obscure bugs and weird performance issues, until I finally decided to eschew it for something that's a little more battle-tested (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4493/commits/b265cbf64af41b8d7d673b746c0a725fbaa37887).

## Description

Here is some important background information from my initial research in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4370#issuecomment-2087620810:

<details>
<summary> PEP 690 - Lazy Imports </summary> 

----

I had the question: "Python style guides generally recommend keeping imports at the top of each module. Isn't lazy-loading incompatible with this? What is the "Pythonic" way to reduce import times?"

It turns out that [**PEP 690 – Lazy Imports**](https://peps.python.org/pep-0690/) (as well as [its rejection](https://discuss.python.org/t/pep-690-lazy-imports-again/19661/26)) explicitly references this dilemma:

> Common Python code style [prefers](https://peps.python.org/pep-0008/#imports) imports at module level, so they don’t have to be repeated within each scope the imported object is used in, and to avoid the inefficiency of repeated execution of the import system at runtime. This means that importing the main module of a program typically results in an immediate cascade of imports of most or all of the modules that may ever be needed by the program.
>
> Consider the example of a Python command line program (CLI) with a number of subcommands. Each subcommand may perform different tasks, requiring the import of different dependencies. But a given invocation of the program will only execute a single subcommand, or possibly none (i.e. if just --help usage info is requested). Top-level eager imports in such a program will result in the import of many modules that will never be used at all; the time spent (possibly compiling and) executing these modules is pure waste.

I couldn't have written a more accurate summary of our current issue.

----

</details>

tl;dr: CLI scripts are uniquely susceptible to import slowdowns, because some options may use expensive packages, but other options may not. In the extreme case, a command like `script --help` shouldn't ever need to load packages like `torch`. Yet, because we don't lazy load our expensive packages, our CLI scripts take quite a long time to load!

----

I wrote a quick test that will check the execution time for `script -h`, then fail if it's over a certain threshold (e.g. 2 seconds). Here are all of the scripts that take longer than 2s on `master`, sorted by duration:

<details>

image

</details>

Here are the times for `script -h` after implementing the fixes in this PR:

<details>

image

</details>

There are probably more gains to be had (by lazy-loading e.g. `nibabel` and `numpy`), but at this point, with script times at ~1s, I think any changes are likely to be diminishing returns. Plus, our usage of `nibabel` and `numpy` are widespread, which would require a lot of tedious updating for every single usage.

----

Much of the slowdown actually came from `inspect.stack()`, so a refactor to remove calls to this function reduced times across our entire CLI, without needing to do any lazy importing at all. (But, lazy importing took care of much of the remaining slowdowns.)

## Linked issues

Fixes #4370.
Fixes https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3077.

### CODE REVIEW

File: contrib/tensorflow/lazy_loader.py
Comment (mguaypaq) [inline contrib/tensorflow/lazy_loader.py:1]: Thankfully this `LazyLoader` class is pretty short, relatively straightforward, and has some good comments. So even though it uses some deeper features (`globals()`, `__getattr__`, `__dir__`, `__dict__`), we should be able to debug hypothetical future problems.

File: testing/cli/test_cli.py [hunk 1]
```diff
@@ -18,12 +19,13 @@
...
+    """
+    Test that SCT's scripts all return error code 2 and show usage descriptions when called with no arguments.
+    Also, ensure that calling the help takes under 2.0 seconds per script.
+    """
+    start_time = time.time()
+    completed_process = subprocess.run([script], capture_output=True)
+    duration = time.time() - start_time
+    assert completed_process.returncode == 2
+    assert b'usage' in completed_process.stderr
+    assert duration < 2.0, f"Expected '{script} -h' to execute in under 2.0s, but took {duration}"
```
Comment (mguaypaq) [inline testing/cli/test_cli.py:31]: If this test becomes flaky in the future, we can probably just raise the threshold, so I'm not too worried about it.

File: spinalcordtoolbox/utils/sys.py [hunk 2]
```diff
@@ -98,16 +100,7 @@ class ANSIColors16(object):
...
-    """Return the first non-`utils.sys` module in the stack (to see where `utils.sys` is being called from)."""
-    for frame in inspect.stack():
-        mod = inspect.getmodule(frame[0])
-        if mod.__file__ != __file__:
-            break
-    return mod
-
-
-def set_loglevel(verbose):
+def set_loglevel(verbose, caller_module_name):
```
Comment (mguaypaq) [inline spinalcordtoolbox/utils/sys.py:103]: I like that this is now explicit, rather than magic.

File: spinalcordtoolbox/utils/shell.py
Comment (mguaypaq) [inline spinalcordtoolbox/utils/shell.py:1]: Given that we only ever call `SCTArgumentParser` with `description` and, occasionally, `epilog`, the whole `update_parent_default` things seems unnecessary. Why not replace `__init__(self, *args, **kwargs)` with `__init__(self, description, epilog=None)`? It's not like we're writing a library and need to be super flexible; we control both the callers and the callee.

File: testing/api/test_utils_shell.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def test_sct_argument_parser(capsys):
 
     # Check that new defaults are set properly
     parser3 = shell.SCTArgumentParser()
-    assert parser3.prog == "test_utils_shell"
+    # assert parser3.prog == "test_utils_shell"  this may actually be `pytest`, `sct_testing`, etc. -- hard to test
```
Comment (mguaypaq) [inline testing/api/test_utils_shell.py:26]: Some parts of this test are clearly written to match the code, rather than being independent requirements/specs. We could probably get rid of those low-value parts.

**Review by mguaypaq — APPROVED** [review]
Thanks for this massive effort, and the research that went into it! It'll be a great quality-of-life improvement.

**Comment by joshuacwnewton** [comment]
Since the RTD failures are upstream (see above issue), this should be good to merge. :)

File: .ci.sh [hunk 4]
```diff
@@ -48,7 +48,8 @@ check_dependencies() {
 
 run_tests() {
   activate_venv_sct
-  pytest testing/api testing/cli
+  # re-run failed tests, just in case something flaky has happened. if no failures, do nothing.
+  pytest testing/api testing/cli || pytest --last-failed --last-failed-no-failures none
```
Comment (joshuacwnewton) [inline .ci.sh:52]: I've added a command here to re-run failed tests, that also [plays nicely with parametrized tests](https://docs.pytest.org/en/latest/how-to/cache.html#rerunning-only-failures-or-failures-first), too. That way, the `2.0s` threshold will only trigger an error if a script takes longer than 2 seconds in back-to-back runs.

File: testing/cli/test_cli.py (hunk 1 above)
Comment (joshuacwnewton) [reply testing/cli/test_cli.py:31]: I've added a command here to re-run failed tests, that also [plays nicely with parametrized tests](https://docs.pytest.org/en/latest/how-to/cache.html#rerunning-only-failures-or-failures-first), too. That way, the `2.0s` threshold will only trigger an error if a script takes longer than 2 seconds in back-to-back runs.

I figure that this is a little more useful than increasing the threshold? Since it keeps a more strict goal for our scripts. (I still want the test to trip if a contributor adds a non-LazyLoaded component that slows down a script.)
